src/js/components/uploadDetachedFiles/uploadDetachedFileMeta.jsx

```diff
@@ -0,0 +1,348 @@
+			dFile.valid = ((start && end) && end.isSameOrAfter(start)) ? true : false;
+
+		this.setState({
+			[file]:dFile,
+			showUploadFilesBox: dFile.valid
+		});
+	}
+
+	updateError(file, header='', description='') {
+		// Show any error that occurs at any point during file upload
+		const state = Object.assign({}, this.state[file], {
+			error: {
+				show: header !='' && description !='',
+				header: header,
+				description: description
+			}
+		});
+		
+		this.setState({
+			[file]: state
+		});
+	}
+
+	uploadFileHelper(local, submission){
+		if(local){
+			return UploadHelper.performDetachedLocalUpload(submission);
+		}
+		return UploadHelper.performDetachedFileUpload(submission);
+	}
+
+	uploadFile() {
+		// upload specified file
+		this.props.setSubmissionState('uploading');
+		let submission = this.props.submission;
+		submission.meta['startDate'] = this.state.detachedAward.startDate.format('DD/MM/YYYY');
+		submission.meta['endDate'] = this.state.detachedAward.endDate.format('DD/MM/YYYY');
+		submission.meta['subTierAgency'] = this.state.agency;
+
+		this.uploadFileHelper(kGlobalConstants.LOCAL == true && !this.isUnmounted, submission)
+			.then((submissionID) => {
+                    // TODO: Remove this when this is eventually tied to user accounts
+                    this.props.setSubmissionId(submissionID);
+					this.checkFileStatus(submissionID);
+					this.props.history.push('/uploaddetachedfiles/'+submissionID);
+					this.props.validate();
+                })
+				.catch((err) => {
+					this.setState({
+						notAllowed: err.httpStatus == 403,
+						errorMessage: err.httpStatus == 403 ? err.message : err.body.message
+					});
+				});
+	}
+
+	checkFileStatus(submissionID) {
+		// callback to check file status
+		GenerateFilesHelper.fetchSubmissionMetadata(submissionID)
+			.then((response) => {
+				if (this.isUnmounted) {
+					return;
+				}
+				const job = Object.assign({}, this.state.jobResults);
+				job.detached_award = response.jobs[0];
+				this.setState({
+					showUploadFilesBox: false,
+					showValidationBox: true,
+					showDatePicker: false,
+					jobResults: job
+				}, () => {
+					this.parseJobStates(response);
+				});				
+			})
+			.catch((err)=>{
+				if(err.status == 400){
+					this.setState({error: 2, submit: false});
+				}
+			});
+	}
+
+	validateSubmission(submissionID){
+		ReviewHelper.validateDetachedSubmission(submissionID)
+				.then((response) => {
+					this.setState({
+						detachedAward: item,
+						validationFinished: true,
+						headerErrors: false,
+						jobResults: response
+					});
+				});
+	}
+
+	parseJobStates(data) {
+		console.log('parse job states')
+		let runCheck = true;
+
+		if (data.jobs[0].job_status == 'failed' || data.jobs[0].job_status == 'invalid') {
+			// don't run the check again if it failed
+			runCheck = false;
+
+			let message = 'Error during D2 validation.';
+
+			if (!data.jobs[0].error_data[0] && data.jobs[0].error_data[0].error_description != '') {
```

Comment:
This is hard to read. Can we move `data.jobs[0]` and `data.jobs[0].error_data[0]` to variables?